sui_types/src/messages.rs

```diff
@@ -311,35 +313,272 @@ where
     }
 }
 
-/// An transaction signed by a client. signature is applied on data.
-/// Any extension to Transaction should add fields to TransactionData, not Transaction.
-// TODO: this should maybe be called ClientSignedTransaction + SignedTransaction -> AuthoritySignedTransaction
-#[derive(Debug, Eq, Clone, Serialize, Deserialize)]
-pub struct Transaction {
+/// A transaction signed by a client, optionally signed by an authority (depending on `S`).
+/// `S` indicates the authority signing state. It can be either empty or signed.
+/// We make the authority signature templated so that `TransactionEnvelope<S>` can be used
+/// universally in the transactions storage in `SuiDataStore`, shared by both authorities
+/// and non-authorities: authorities store signed transactions, while non-authorities
+/// store unsigned transactions.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(remote = "TransactionEnvelope")]
+pub struct TransactionEnvelope<S> {
     // Deserialization sets this to "false"
     #[serde(skip)]
     pub is_checked: bool,
 
     pub data: TransactionData,
-    pub signature: Signature,
+    /// tx_signature is signed by the transaction sender, applied on `data`.
+    pub tx_signature: Signature,
+    /// auth_signature, if available, is signed by an authority, applied on `data`.
+    pub auth_signature: S,
+    // Note: If any new field is added here, make sure the Hash and PartialEq
+    // implementation are adjusted to include that new field (unless the new field
+    // does not participate in the hash and comparison).
 }
-/* TODO: check this legacy static check is not needed as an invariant
-         elsewhere. The tests pass but invariant checks without docs
-         should not be removed without a note to check.
 
-const_assert_eq!(
-    size_of::<TransactionData>() + size_of::<Signature>(),
-    size_of::<Transaction>()
-);
+impl<S> TransactionEnvelope<S> {
```

Comment:
OK, so now we see that there are no constraints on the transaction envelope's signature contents. This is worrisome for two reasons:
- if the type is unconstrained, there are no methods acting on it, except to ferry the data around. Is the signature data actually going to be used for something in the future, do we have a clear idea on what that should be, and have we equipped the future developer going over this with enough info to deal with this? If not, could we literally have code comments pointing to a description of the way this embedded data is going to evolve?
- anyone can now chuck whatever they want in there, rather than the two types we have in mind. In order to limit the parameters to a specific (but extensible) set of traits, we can use [the "sealed trait" pattern](https://rust-lang.github.io/api-guidelines/future-proofing.html), by having an empty (aka marker) trait with private access, and making `XXAuhoritySIgnInfo` instances of it. The trait would be required on the line this current comment points at.

Reply:
> For this reason, we cannot implement signature verification on it.

Naturally. But the point is that the reader of this code should ideally have a mental model of two specific instantiations of the template, and will mentally place them in very different places in the state machine, as one carries an empty signature (and is untrusted), and one carries a signature that should be validated (and can be validated).

The sealed trait should help the reader in creating this mental model. I guess what I'm saying is a few comments on the template itself: "here are the instances of this type we expect you to encounter" would also help.